src/daft-catalog/src/bindings.rs

```diff
@@ -1,46 +1,87 @@
-use std::collections::HashMap;
-
-/// Bindings
-///
-/// Notes:
-///  - Currently using Clone because all references are arcs.
-///  - This is intentionally lightweight and everything is exact-case.
-///  - All APIs are non-fallible because callers determine what is an error.
-///  - It does not necessarily have map semantics.
-///  - Callers are responsible for case-normalization hence String, &str.
-///  - Intentionally using String and &str rather than Into and AsRef.
+use std::collections::{HashMap, HashSet};
+
+/// Bindings are stored case-normalized with the case-preserved bindings.
 #[derive(Debug)]
-pub struct Bindings<T>(HashMap<String, T>);
+pub struct Bindings<T> {
+    /// case-preserved bindings (default)
+    bindings: HashMap<String, T>,
+    /// case-normalized binding aliases for insensitive lookups.
+    aliases: HashMap<String, HashSet<String>>,
+}
+
+/// Bind mode is for lvalue handling which is currently out of scope.
+#[derive(Debug, Clone, Copy)]
+pub enum BindMode {
+    Preserve,
+    Normalize,
+}
 
+/// Find mode is for rvalue handling.
+#[derive(Debug, Clone, Copy)]
+pub enum FindMode {
+    Insensitive,
+    Sensitive,
+}
+
+/// Bindings implements "bind" and "find" for handling various identifier modes.
 impl<T> Bindings<T> {
-    /// Creates an empty catalog provider.
+    /// Creates an empty bindings object.
     pub fn empty() -> Self {
-        Self(HashMap::new())
+        Self {
+            bindings: HashMap::new(),
+            aliases: HashMap::new(),
+        }
     }
 
-    /// Inserts a new binding with ownership.
-    pub fn insert(&mut self, name: String, object: T) {
-        self.0.insert(name, object);
+    /// Bind an object to the name (lvalue), overwriting any existing binding.
+    pub fn bind(&mut self, name: String, object: T) {
+        self.bindings.insert(name.clone(), object);
+        self.aliases.entry(alias(&name)).or_default().insert(name);
     }
 
-    /// Removes the binding if it exists.
-    pub fn remove(&mut self, name: &str) {
-        self.0.remove(name);
+    /// Find all object by name (rvalue).
+    pub fn find(&self, name: &str, mode: FindMode) -> Vec<&T> {
+        match mode {
+            FindMode::Insensitive => self.find_aliased(name),
+            FindMode::Sensitive => self.find_binding(name),
+        }
+    }
+    /// Finds this binding by exact case.
+    #[inline]
+    fn find_binding(&self, name: &str) -> Vec<&T> {
+        self.bindings
+            .get(name)
+            .map(|binding| vec![binding])
+            .unwrap_or_default()
+    }
+
+    /// Finds all bindings mapped to this alias.
+    #[inline]
```

Comment:
what's the justification for `#[inline]`ing these?

Reply:
private and only used in one place, was to simplify case branches, it's superfluous and can be removed